paper/src/main/java/com/untamedears/realisticbiomes/growth/NMSFeatureGrower.java

```diff
@@ -0,0 +1,188 @@
+package com.untamedears.realisticbiomes.growth;
+
+import java.util.Iterator;
+import java.util.Random;
+
+import org.bukkit.Material;
+import org.bukkit.block.Block;
+import org.bukkit.block.data.type.Sapling;
+import org.bukkit.craftbukkit.v1_18_R2.block.CraftBlock;
+
+import com.untamedears.realisticbiomes.model.Plant;
+
+import net.minecraft.core.BlockPos;
+import net.minecraft.data.worldgen.features.TreeFeatures;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.tags.BlockTags;
+import net.minecraft.world.level.LevelAccessor;
+import net.minecraft.world.level.block.Blocks;
+import net.minecraft.world.level.block.grower.AbstractMegaTreeGrower;
+import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.chunk.ChunkGenerator;
+import net.minecraft.world.level.levelgen.feature.ConfiguredFeature;
+
+public class NMSFeatureGrower extends AgeableGrower {
+
+	public NMSFeatureGrower(Material material) {
+		super(material, 1, 1);
+	}
+
+	private boolean isMega;
+
+
+
+
+	@Override
+	public int getStage(Plant plant) {
+		Block block = plant.getLocation().getBlock();
+		if (block.getType() != this.material) {
+			return -1;
+		}
+		return 0;
+	}
+
+	@Override
+	public boolean setStage(Plant plant, int stage) {
+		if (stage < 1) {
+			return true;
+		}
+		Block block = plant.getLocation().getBlock();
+		// Re-Read the block data to make sure it is up to date
+		if (!(block.getBlockData() instanceof Sapling)) {
+			return true;
+		} 
+		
+		
+		CraftBlock cBlock = (CraftBlock) block;
+		BlockPos pos = cBlock.getPosition();
+		ServerLevel spWorld = cBlock.getCraftWorld().getHandle();
+		BlockState spBlock = cBlock.getNMS();
+
+		this.growTreeFeature(spWorld, spWorld.getChunkSource().getGenerator(), pos, spBlock, new Random());
+		return true;
+	}
+	
+	public boolean growTreeFeature(
+			ServerLevel world, 
+			ChunkGenerator chunkGenerator, 
+			BlockPos pos, 
+			BlockState state, 
+			Random random) {
+		isMega = false;
+		if(placeMegaTree(world, chunkGenerator, pos, state, random)) {return true;};
+		if(isMega) return false; // fail tree growth  here if it's a mega tree and it hasn't returned true above
+		ConfiguredFeature<?, ?> feature;
+		ConfiguredFeature<?, ?> featureSmall;
+		//feature holds the default Minecraft tree and featureSmall holds a version with randomHeight set to 0 so that it will grow at minimum height
+		switch(state.getBukkitMaterial()) {
+		case JUNGLE_SAPLING:
+			feature = TreeFeatures.JUNGLE_TREE_NO_VINE.value();
+			featureSmall = RBFeatures.SHORT_JUNGLE.value();
+			break;
+		case SPRUCE_SAPLING:
+			feature = TreeFeatures.SPRUCE.value();
+			featureSmall = RBFeatures.SHORT_SPRUCE.value();
+			break;
+		case OAK_SAPLING:
+			feature = this.hasFlowers(world, pos) ? TreeFeatures.OAK_BEES_005.value() : TreeFeatures.OAK.value();
```

Comment:
small note but does NMS not contain a method to do this for us?

Reply:
Specifically checking nearby for flowers

Reply:
it does but it's private so I copied it over